tiledb/sm/array_schema/array_schema.cc

```diff
@@ -545,95 +586,150 @@ Status ArraySchema::drop_attribute(const std::string& attr_name) {
+    cell_validity_filters = cell_validity_filters_deserialized.value();
   }
 
   // Load domain
-  auto&& [st_domain, deserialized_domain]{
-      Domain::deserialize(buff, version_, cell_order_, tile_order_)};
+  auto&& [st_domain, deserialized_domain]{Domain::deserialize(
+      buff, version, Layout(cell_order), Layout(tile_order))};
   if (!st_domain.ok()) {
-    return st_domain;
+    throw std::runtime_error(
+        "[ArraySchema::deserialize] Error: Cannot deserialize domain.");
   }
-
-  domain_ = deserialized_domain.value();
+  shared_ptr<Domain> domain = deserialized_domain.value();
 
   // Load attributes
+  std::vector<shared_ptr<const Attribute>> attributes;
+  std::unordered_map<std::string, const Attribute*> attribute_map;
   uint32_t attribute_num;
-  RETURN_NOT_OK(buff->read(&attribute_num, sizeof(uint32_t)));
+  st = buff->read(&attribute_num, sizeof(uint32_t));
+  if (!st.ok())
+    throw std::runtime_error(
+        "[ArraySchema::deserialize] Error: Cannot load attribute_num.");
   for (uint32_t i = 0; i < attribute_num; ++i) {
-    auto&& [st_attr, attr]{Attribute::deserialize(buff, version_)};
+    auto&& [st_attr, attr]{Attribute::deserialize(buff, version)};
     if (!st_attr.ok()) {
-      return st_attr;
+      throw std::runtime_error(
+          "[ArraySchema::deserialize] Error: Cannot deserialize attributes.");
     }
 
-    attributes_.emplace_back(
-        tdb::make_shared<Attribute>(HERE(), move(attr.value())));
-    auto a = attributes_.back().get();
-    attribute_map_[a->name()] = a;
+    attributes.emplace_back(make_shared<Attribute>(HERE(), move(attr.value())));
+    auto a = attributes.back().get();
+    attribute_map[a->name()] = a;
   }
 
   // Create dimension map
-  auto dim_num = domain()->dim_num();
+  std::unordered_map<std::string, shared_ptr<const Dimension>> dim_map;
+  auto dim_num = domain->dim_num();
   for (unsigned d = 0; d < dim_num; ++d) {
-    auto dim = dimension(d);
-    dim_map_[dim->name()] = dim;
+    auto dim = domain->dimension(d);
+    dim_map[dim->name()] = dim;
   }
 
-  // Initialize the rest of the object members
-  RETURN_NOT_OK(init());
+  // Populate timestamp range
+  std::pair<uint64_t, uint64_t> timestamp_range;
+  st = utils::parse::get_timestamp_range(uri, &timestamp_range);
+  if (!st.ok())
+    throw std::runtime_error(
+        "[ArraySchema::deserialize] Error: Cannot get timestamp_range.");
+
+  // Get schema name
+  std::string name = uri.last_path_part();
 
   // Success
-  return Status::Ok();
+  return make_shared<ArraySchema>(
+      HERE(),
+      allows_dups,
+      URI(),
+      (ArrayType)array_type,
+      attribute_map,
+      attributes,
+      capacity,
+      (Layout)cell_order,
+      cell_var_filters.value(),
+      cell_validity_filters,
+      coords_filters.value(),
+      dim_map,
+      domain,
+      (Layout)tile_order,
+      version,
+      timestamp_range,
+      uri,
+      name);
```

Comment:
We shouldn't have C-style casts at all, and it would be better not to be have any casts here at all. Better to deal with any casting next to the code where the value is initially read. The cast should happen after validation, because only then do you know that the datum is valid as the type it's cast to.